### DISCUSSION TITLE
Cannot resolve dependencies of MavenProject -> org.junit

### DISCUSSION THREAD

**Discussion by JamesPoli** [eclipse-tycho/tycho#2285] [Q&A]
Hello,
I'm upgrading our Eclipse plug-ins from the Eclipse 2021-09 building with Java 11 to Eclipse 2023-03 building with Java 17.  The original Maven 3.8.2 build used Tycho 2.2.0; now, I'm using 3.0.4.  I am stuck on a transitive Tycho dependency on an old version of Junit.  I included JUnit dependency in the parent pom, and it still fails. I've read previous discussions but am still trying to find a solution.  I would appreciate any help.  
Regards, Jim

--- **BUILD OUTPUT** ---
mvn clean verify -X
...
Apache Maven 3.8.2 (ea98e05a04480131370aa0c110b8c54cf726c06f)
Maven home: C:\THIRD PARTY TOOLS\MAVEN\apache-maven-3.8.2
Java version: 17.0.6, vendor: Azul Systems, Inc., runtime: C:\Program Files\Zulu\zulu17.40.20-sa-jdk17.0.6-win_x64
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
...
[INFO]  org.eclipse.equinox.security.linux.x86_64 1.1.500.v20210125-1508
[INFO] [main] Resolving test dependencies of MavenProject: IFABSDE:sas.tools.ifabs.playpen:17.0.0 @ C:\IFABSDE\RELTOOLSSUP-469\2023-03\work\eclipse-plugins\com.sas.tools.ifabsde.parent\sas.tools.ifabs.playpen\pom.xml
[DEBUG] No solution found because the problem is unsatisfiable.: [Unable to satisfy dependency from tycho-extra-1679673309392 0.0.0.1679673309392 to org.eclipse.equinox.p2.iu; org.junit [3.8.2,3.9.0).; Unable to satisfy dependency from tycho-1679673309438 0.0.0.1679673309438 to org.eclipse.equinox.p2.iu; org.junit [3.8.2,3.9.0).; No solution found because the problem is unsatisfiable.]
[INFO] {osgi.os=win32, osgi.ws=win32, org.eclipse.update.install.features=true, osgi.arch=x86_64}
**[ERROR] Cannot resolve project dependencies:
[ERROR]   You requested to install 'org.eclipse.equinox.p2.iu; org.junit [3.8.2,3.9.0)' but it could not be found
[ERROR]**
[ERROR] See https://wiki.eclipse.org/Tycho/Dependency_Resolution_Troubleshooting for help.
[ERROR] Cannot resolve dependencies of MavenProject: IFABSDE:sas.tools.ifabs.playpen:17.0.0 @ C:\IFABSDE\RELTOOLSSUP-469\2023-03\work\eclipse-plugins\com.sas.tools.ifabsde.parent\sas.tools.ifabs.playpen\pom.xml: See log for details -> [Help 1]
org.apache.maven.MavenExecutionException: Cannot resolve dependencies of MavenProject: IFABSDE:sas.tools.ifabs.playpen:17.0.0 @ C:\IFABSDE\RELTOOLSSUP-469\2023-03\work\eclipse-plugins\com.sas.tools.ifabsde.parent\sas.tools.ifabs.playpen\pom.xml

**Comment by JamesPoli** [comment]
Is a p2 repository required for the JUnit dependency?  I have a typical Maven 2 repository dependency, but it doesn't appear to use it.

<dependency>
            <groupId>junit</groupId>
            <artifactId>junit</artifactId>
            <version>3.8.2</version>
            <scope>test</scope>
</dependency>

**Reply by laeubi** [reply]
The best is if you provide an example.

**Comment by JamesPoli** [comment]
**I've attached the parent pom and the failing child pom as zip files.  Jim**

**java -version**
openjdk version "17.0.6" 2023-01-17 LTS
OpenJDK Runtime Environment Zulu17.40+20-SA (build 17.0.6+10-LTS)
OpenJDK 64-Bit Server VM Zulu17.40+20-SA (build 17.0.6+10-LTS, mixed mode, sharing)

**mvn -version**
Apache Maven 3.8.2 (ea98e05a04480131370aa0c110b8c54cf726c06f)
Maven home: C:\THIRD PARTY TOOLS\MAVEN\apache-maven-3.8.2
Java version: 17.0.6, vendor: Azul Systems, Inc., runtime: C:\Program Files\Zulu\zulu17.40.20-sa-jdk17.0.6-win_x64
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

**mvn clean package**
[INFO] Scanning for projects...
[WARNING] Using JavaSE-17 to fulfill requested profile of JavaSE-11 this might lead to faulty dependency resolution, consider define a suitable JDK in the toolchains.xml
[INFO] ### Using TychoRepositoryTransport for remote P2 access ###
[INFO]     Cache location:         C:\Users\japoli\.m2\repository\.cache\tycho
[INFO]     Transport mode:         online
[INFO]     Update mode:            cache first
[INFO]     Minimum cache duration: 60 minutes
[INFO]       (you can configure this with -Dtycho.p2.transport.min-cache-minutes=<desired minimum cache duration>)
...
[INFO] Resolving dependencies of MavenProject: IFABSDE:sas.tools.ifabs.playpen:17.0.0 @ C:\IFABSDE\RELTOOLSSUP-469\2023-03\work\eclipse-plugins\com.sas.tools.ifabsde.parent\sas.tools.ifabs.playpen\pom.xml
[INFO] Resolving test dependencies of MavenProject: IFABSDE:sas.tools.ifabs.playpen:17.0.0 @ C:\IFABSDE\RELTOOLSSUP-469\2023-03\work\eclipse-plugins\com.sas.tools.ifabsde.parent\sas.tools.ifabs.playpen\pom.xml
[INFO] {osgi.os=win32, osgi.ws=win32, org.eclipse.update.install.features=true, osgi.arch=x86_64}
[ERROR] Cannot resolve project dependencies:
[ERROR]   You requested to install 'org.eclipse.equinox.p2.iu; org.junit [3.8.2,3.9.0)' but it could not be found
[ERROR]
[ERROR] See https://wiki.eclipse.org/Tycho/Dependency_Resolution_Troubleshooting for help.
[ERROR] Cannot resolve dependencies of MavenProject: IFABSDE:sas.tools.ifabs.playpen:17.0.0 @ C:\IFABSDE\RELTOOLSSUP-469\2023-03\work\eclipse-plugins\com.sas.tools.ifabsde.parent\sas.tools.ifabs.playpen\pom.xml: See log for details -> [Help 1]
[ERROR]
[pom-parent.zip](https://github.com/eclipse-tycho/tycho/files/11079298/pom-parent.zip)

[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[pom-child.zip](https://github.com/eclipse-tycho/tycho/files/11079299/pom-child.zip)

[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MavenExecutionException

**Reply by JamesPoli** [reply]
Sorry in advance about the location of the zips in the message.  Jim

**Comment by JamesPoli** [comment]
I need more ideas on this one.  FWIW, the failure comes from the one module containing a JUnit test and many plug-in dependencies.  Is there a way to force the inclusion of JUnit to satisfy the error?  Or force Maven to exclude it.  I like the Eclipse Tycho plug-in to build on the command line.  Of course, it builds and runs in Eclipse 2023-03 using the PDE.

**Comment by laeubi** [comment]
@JamesPoli The parent pom shows that your are mixing different releases, what might or might not work out well with a lot of custom configuration and child modules, and the child pom you presented here does not show the important parts but I can see that it is actually a `eclipse-plugin`, so maybe it contains test source folders that Tycho now tries to execute.

To narrow things down you can take a look here, where it explains the different setups and links to working demo examples:

- https://tycho.eclipseprojects.io/doc/master/TestingBundles.html

Also please try out the current [tycho snapshot](https://github.com/eclipse/tycho/wiki#getting-tycho-snapshots) build (`4.0.0-SNAPSHOT`) if that maybe improve things.

The you can try to provide an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue, beside that, if it is crucial to your business you can find dedicated support options for Tycho here:
- https://github.com/eclipse-tycho/tycho#getting-support

**Comment by JamesPoli** [comment]
Thanks, @laeubi.  You are likely right about mixing Eclipse releases as a possible cause of the issue.  I added Eclipse 2021-09 because I require the Eclipse CVS Client to develop older SAS products.  

I added an option to the Maven build, i.e., -Dtycho.debug.resolver=true, to help debug the issue.  For the module in question, I do see the following:
...
[DEBUG] Extra Requirements:
   org.eclipse.equinox.p2.iu; org.junit [3.8.2,3.9.0)
...
Unfortunately, I need help seeing why Maven is including it transitively.  

Thanks for the links.  Due to time constraints, I'm planning to use the PDE to build the plug-ins.  I'll come back to Maven afterward.  I inherited the plug-in code, which only used the PDE for development.  I added the Maven Tycho support myself, but I need to become an expert; I'm finding out now that I need to upgrade Eclipse to 2023-03.  

Regards, Jim
